### ISSUE TITLE
SwiftUI - Global accent color override

### ISSUE THREAD

**Issue by esreli** [mixpanel/mixpanel-swift#522]
## Environment

Xcode Version 13.2.1 (13C100)
Apple Swift version 5.5.2 (swiftlang-1300.0.47.5 clang-1300.0.29.30)
SPM Resolved Dependency:

```txt
      {
        "package": "Mixpanel",
        "repositoryURL": "https://github.com/mixpanel/mixpanel-swift",
        "state": {
          "branch": "master",
          "revision": "8f6fcb1876f48ef4e7794c66719b23838c990410",
          "version": null
        }
      }
```

## Issue

I'm building a SwiftUI iOS app with a min deployment of iOS 14+. Because of this I'm using the SwiftUI types introduced in iOS 14 like `App`, `Scene`, etc. As such, I'm not notified that the app has launched through the app or scene delegate as is the case for `UIKit` apps. With this "pure" SwiftUI approach, all initializations are typically done within the `App`'s init function. 

With the initialization of Mixpanel, all of a sudden the project's Global accent color (contained by an Asset Catalog) is reverted to the default system blue.

```swift
@main
struct MyApp: App {
        
    init() {
        // Initialize Mixpanel
        Mixpanel.initialize(token: .mixpanelToken)
        #if DEBUG
        Mixpanel.mainInstance().loggingEnabled = true
        #endif
    }
    
    var body: some Scene {
        WindowGroup {
            MyAppView()
        }
    }
}
```

I believe this is an issue with the initialization API because when I comment out the line...

```swift
// Mixpanel.initialize(token: .mixpanelToken)
``` 

...the app global accent color works as expected.

### What I've tried to remedy the issue

- Changing the Target's **Global Accent Color Name** & the corresponding `.xcasset` item from the default `"AccentColor"` to a non-standard name.
- Checking the build log to see if other asset catalogs were unexpectedly compiled into the binary
- Forking the repo & updating the Xcode project settings

### Current Workaround

I've created a lazily built reference to `Mixpanel.mainInstance()` in which I initialize the token. This allows the Mixpanel initialize token code to be ran on the first time the app fires off the first track event. 

```swift 
lazy var instance: MixpanelInstance = {
    // Mixpanel
    Mixpanel.initialize(token: .mixpanelToken)
    #if DEBUG
    Mixpanel.mainInstance().loggingEnabled = true
    #endif
    return Mixpanel.mainInstance()
}()
```

This isn't ideal because any tracking associated with initializing Mixpanel is deferred until the first actively tracked event leaving a gap between launch and the first event in unknown territory.

---
Do you have any idea what's going on here? Thanks!

**Comment by jaredmixpanel** [comment]
@esreli can you provide a simple, example project that consistently reproduces this issue? I've managed to reproduce a few times, somewhat randomly, but more often than not, the issue is not reproducible.

**Comment by esreli** [comment]
Hi @jaredmixpanel, thanks for looking into this issue. After this sprint I'll try to set aside some time to publish a consistent repro case.

**Comment by esreli** [comment]
Hi @jaredmixpanel. I've reproduced the bug and pushed the code to this repository: [esreli/mixpanel-bug](https://github.com/esreli/mixpanel-bug).

Here's another interesting thing that I've learned along the way. If I were to initialize Mixpanel in the highest level [`View`](https://github.com/esreli/mixpanel-bug/blob/main/Shared/ContentView.swift#L12) instead of initializing Mixpanel within the [`App`](https://github.com/esreli/mixpanel-bug/blob/main/Shared/Mixpanel_SwiftUI_AccentColor_BugApp.swift#L18) then the problem seems to resolve itself. 

Thanks in advance for looking into this for me!

**Comment by jaredmixpanel** [comment]
@esreli thanks for the example project. I still find the behavior very random. I've attached a long, boring, low resolution (sorry!) screen recording below, so you can see my results are very inconsistent. At the moment, I'm inclined to think this is a SwiftUI/Xcode kind of thing and that Mixpanel just so happens to be the code that is in your init(), but that some other code may cause similar issues.

https://user-images.githubusercontent.com/10504508/158898751-9a33c130-d451-4e35-853b-3c0a0f13543b.mp4

**Comment by loryhuz** [comment]
Hi @esreli I've still the issue, my whole app turns blue when I init Mixpanel

**Comment by vulgur** [comment]
same here, and the issue happens every time, not randomly.

**Comment by jaredmixpanel** [comment]
From what I can gather this can happen whenever some UIKit code gets executed "too early".

https://forums.developer.apple.com/forums/thread/691336

https://stackoverflow.com/questions/71213275/wrong-accentcolor-showing-color-ignored

https://stackoverflow.com/questions/69175418/ios15-xcode-13-global-accent-color-not-working

Try moving your Mixpanel initialization code to AppDelegate or SceneDelegate or...

```
    var body: some Scene {
        WindowGroup {}
        .onChange(of: scenePhase) { _, newPhase in
            if newPhase == .active {
                Mixpanel.initialize(token: "YOUR_TOKEN", trackAutomaticEvents: true)
            }
        }
    }
```

Seems like another option is to manually specify the accent color on the WindowGroup...

```
@main
struct MyApp: App {
    var body: some Scene {
        WindowGroup {
            myMainView
                .accentColor(MyGlobalAppearance.myAccentColor)
        }
    }
}
```

**Comment by vulgur** [comment]
@jaredmixpanel 
I tried both of your suggestions.
Putting initialization in onChange works, but setting accentColor still doesn't work.

**Comment by vulgur** [comment]
> [@vulgur](https://github.com/vulgur) Sorry to hear your still having issues. So have you tried...
> 
> * [ ]  1. Initializing Mixpanel in the AppDelegate's `applicationDidBecomeActive(_:)`[ ]  2. Or if you are using Scenes, in the SceneDelgate's `sceneDidBecomeActive(_:)`[ ]  3. Initializing Mixpanel in the `.onChange(of: scenePhase)`[ ]  4. Manually setting the `.accentColor()`on the main view in the WindowGroup[ ]  5. Manually setting the window's `.tintColor` in the SceneDelegate as mentioned here: https://stackoverflow.com/a/73461232[ ]  6. Combining various permutations of the above like 3 + 4 or 2 + 5, etc..
> 
> I would suggest continue experimenting with initializing Mixpanel at various points of your application / scene lifecycle and manually setting the `accentColor` and `tintColor` in code. The Mixpanel SDK doesn't touch accent color but it does depend on a few UIKit API's and apparently if your app calls some UIKit API's too early, before calling WindowGroup it seems, this issue can occur as discussed in the links provided.

I am not using AppDelegate and SceneDelegate, so I have not tried 1, 2 and 5.
As I mentioned above, 3 works for me, but 4 don't works.

Thanks for your explanation and suggestions.

**Comment by jaredmixpanel** [comment]
@alpennec ah, bummer... 😞 okay, i will look into that.

**Comment by Martma12** [comment]
@alpennec do we have any fix for this yet?

**Comment by alpennec** [comment]
@Martma12 I don't know, I thought it had been fixed by the people working at Mixpanel (@jaredmixpanel @santigracia  @rahul-mixpanel @ketanmixpanel). I'll try to test with an empty project if I have some time in the coming days. You can still contribute to the open source SDK if you find the reason of the issue and open a PR to fix it.

**Comment by Martma12** [comment]
@alpennec I apologize I thought you work for Mixpanel.

**Comment by santigracia** [comment]
@alpennec and @Martma12 I'm happy to confirm that [version 6.3.0 ](https://github.com/mixpanel/mixpanel-swift/releases/tag/v6.3.0) has been released a few hours ago with a fix for this issue. Thank you so much for reporting it and for providing all the details. This helped me build a small project to reproduce it.  [https://github.com/mixpanel/mixpanel-swift/releases/tag/v6.3.0](https://github.com/mixpanel/mixpanel-swift/releases/tag/v6.3.0)

For context, the issue was caused by a race condition when initializing with automaticEvents true using SwiftUI and was mainly affecting accentColor only on first app run (sometimes second app run too). The fix I submitted defers automatic events initialization to prevent the race condition.

Feel free to give it a try using 6.3.0 and let us know if the issue is fully resolved also on your end.

Best,
Santi Gracia - Mixpanel

**Comment by alpennec** [comment]
Hello @santigracia,

I just tested again in a project that uses Mixpanel, and the issue does not seem to be fixed with 6.3.0.

Here is how I initialise in my App.init:

```
@main
struct SkipperApp: App {
    init() {
          Mixpanel.initialize(
               token: "e93085c87c...",
               trackAutomaticEvents: false
          )
    }

     var body: some Scene {
        WindowGroup {
            Button("This is a button") { }
                .buttonStyle(.borderedProminent)
            }
     }
}
```

Sometimes the button gets the correct accentColor, sometimes not.

Maybe there is a conflict somewhere else with something I'm doing in my App init. But it only occurs when I initialize the Mixpanel library alongside my other code.

Axel

**Comment by alpennec** [comment]
Hello again @jaredmixpanel @santigracia @rahul-mixpanel @ketanmixpanel,

I just checked my own fork with a fix I added few months ago. This fork is working correctly with my code. But if I use your repository, it's not working.

Your SDK accesses the UIScreen.main.bounds.size from the [AutomaticProperties.properties](https://github.com/mixpanel/mixpanel-swift/blob/5ecded282e1468a5b6ac60e16569cc074f814db3/Sources/AutomaticProperties.swift#L27). I disabled that in my fork (see my other reply here: https://github.com/mixpanel/mixpanel-swift/pull/682#issuecomment-3669915133).

See my post on X about this https://x.com/alpennec/status/2001629548383998291?s=20 or @sebjvidal post about this SwiftUI issue https://x.com/SebJVidal/status/2039637672914829348?s=20

Can you please update your library to make this finally work correctly without interfering with our app's accent color?

Thank you,
Axel

**Comment by ketanmixpanel** [comment]
@alpennec Thanks for the feedback! On my end, the sample app I tested with showed the issue as fixed, so it would really help if you could share the sample app where you're still able to reproduce it. That way I can dig into the difference and figure out what's still going wrong.

**Comment by alpennec** [comment]
@ketanmixpanel thanks for the follow up. I re created a project from scratch in order to identify the source of the issue.

Here is what I have in my App for a completely new project (the other code is unchanged, except importing the Mixpanel package).

```
import CoreData
import Mixpanel
import SwiftUI

@main
struct MixpanelTestApp: App {
    init() {
        // No issue if only this line but issue if both Mixpanel.initialize and NSManagedObjectModel.mergedModel are used
        Mixpanel.initialize(
            token: "",
            trackAutomaticEvents: true
        )
        
        // No issue if only this line but issue if both Mixpanel.initialize and NSManagedObjectModel.mergedModel are used
        let model: NSManagedObjectModel? = NSManagedObjectModel.mergedModel(from: nil)
    }
    
    var body: some Scene {
        WindowGroup {
            ContentView()
        }
    }
}
```

If I comment or remove the line that accesses the model, the accent color is correct.
If I comment or remove the Mixpanel init, the accent is correct.
If in the Mixpanel SDK [I comment the line that accesses](https://github.com/alpennec/mixpanel-swift/blob/4388d92390d004dc589130726652d37e82ab391d/Sources/AutomaticProperties.swift#L27) the `UIScreen.main.bounds.size`, the accent color is correct.

You can try with my fork: https://github.com/alpennec/mixpanel-swift and see the accent color is always correct no matter what happens in the App init.

Thanks,
Axel

**Comment by ketanmixpanel** [comment]
Thanks @alpennec, I am able to reproduce the issue with the code you shared. Let me dig into it and get back to you.

**Comment by ketanmixpanel** [comment]
@alpennec I have worked on a fix for this issue, available in the branch `fix/global-accent-color-issue`. Could you give it a try on your side and let me know if it resolves the issue for you?

**Comment by ketanmixpanel** [comment]
I've pushed a small fix, feel free to pull the latest and verify on your end.

**Comment by alpennec** [comment]
@ketanmixpanel I just tested your branch with my real project, and the issue seems to be fixed at first sight. Using the current release (6.4.0), the accent color is not passed to Views in one of my packages, but with your fix, it is. So it seems to be working.

**Comment by ketanmixpanel** [comment]
Thanks for confirming @alpennec. I will update it here when it's released.

**Comment by alpennec** [comment]
@ketanmixpanel any update on this? Thank you!

**Comment by tylerjroach** [comment]
Hi @alpennec Sorry for the delay, we will try to push out a release shortly with the official fix. Thanks again for your help!

**Comment by ketanmixpanel** [comment]
Hi @alpennec , Apologies for the delay in releasing this update. The changes are now available in version [6.4.1](https://github.com/mixpanel/mixpanel-swift/releases/tag/6.4.1). Thank you for your help and patience throughout the process.

**Comment by alpennec** [comment]
Thanks for letting me know! I'll update my dependencies then.
